Server Side Field Validation

kevindelgado opened the pull request:
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Performs strict server side schema validation via the `fieldValidation=[Strict,Warn,Ignore]` query parameter. 

This change is dependent on `liggitt/json-stdlib` and so this PR is opened on this fork in order to begin collecting feedback on it until `liggitt/json-stdlib` is merged into upstream k/k.

It introduces the `fieldValidation` query parameter that when set to `Strict` causes requests to error when unknown fields are present, when set to `Warn` succeeds the request but returns a warning, and when set to `Ignore` it ignores unknown fields.

For create/update requests, we use strict json unmarshalling to determine the unknown fields and error/warn based on those fields.

For JSON Patch requests (i.e. CRDs), we use the prune mechanism in the customresource handler to resolve any unknown fields.

For SMP Patch request, we use the the unstructured converter to resolve unknown fields.

Apply Patch requests already require strict schemas and will error on unknown fields.

#### Which issue(s) this PR fixes:
First step in solving #39434 and #5889 

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Performs strict server side schema validation requests via the `fieldValidation=[Strict,Warn,Ignore]` query parameter.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
[KEP-2885](https://github.com/kubernetes/enhancements/issues/2885)
<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
[KEP-2885](https://github.com/kubernetes/enhancements/issues/2885)
```

apelisse commented on staging/src/k8s.io/apimachinery/pkg/runtime/converter.go:
```diff
@@ -358,41 +388,64 @@ func pointerFromUnstructured(sv, dv reflect.Value) error {
+		allStrictErrs := make([]error, len(keys))
+		for i, unknownField := range keys {
+			allStrictErrs[i] = fmt.Errorf("unknown field: %s", unknownField.String())
+			i++
```
nit but you can drop the `i` here too

apelisse commented on staging/src/k8s.io/apimachinery/pkg/runtime/converter.go:
```diff
@@ -358,41 +388,64 @@ func pointerFromUnstructured(sv, dv reflect.Value) error {
 		} else {
+			// delete from the source's keys
+			deleteFromKeys(fieldInfo.name, &keys)
+			delete(fields, fieldInfo.name)
```
It's a little sad that we have to change all the function signatures to carry this field that is used here. Let's see if we can do better later.

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/apiserver_test.go:
```diff
@@ -4388,6 +4388,116 @@ func (storage *SimpleRESTStorageWithDeleteCollection) DeleteCollection(ctx conte
+		errContains string
+	}{
+
+		{name: "post-unknown-strict-validation", path: "/namespaces/default/simples", verb: "POST", data: []byte(`{"kind":"Simple","apiVersion":"test.group/version","metadata":{"creationTimestamp":null},"other":"bar","unknown":"baz"}`), queryParams: strictFieldValidation, errContains: strictDecoderErr},
```
Can we warp these lines that are hard to read?

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/apiserver_test.go:
```diff
@@ -4388,6 +4388,116 @@ func (storage *SimpleRESTStorageWithDeleteCollection) DeleteCollection(ctx conte
+		{name: "post-unknown-ignore-validation", path: "/namespaces/default/simples", verb: "POST", data: []byte(`{"kind":"Simple","apiVersion":"test.group/version","metadata":{"creationTimestamp":null},"other":"bar","unknown":"baz"}`)},
+		{name: "put-unknown-ignore-validation", path: "/namespaces/default/simples/id", verb: "PUT", data: []byte(`{"kind": "Simple", "apiVersion": "test.group/version", "metadata": {"name": "id", "creationTimestamp": null}, "other": "bar", "unknown": "baz"}`)},
+		{name: "post-unknown-strict-vaidation-json", path: "/namespaces/default/simples", verb: "POST", data: []byte(`{"kind":"Simple","apiVersion":"test.group/version","metadata":{"creationTimestamp":null},"other":"bar","unknown":"baz"}`), queryParams: strictFieldValidation, errContains: strictDecoderErr, contentType: runtime.ContentTypeJSON},
+		{name: "post-unknown-strict-validation-protobuf", path: "/namespaces/default/simples", verb: "POST", data: []byte(`{"kind":"Simple","apiVersion":"test.group/version","metadata":{"creationTimestamp":null},"other":"bar","unknown":"baz"}`), queryParams: strictFieldValidation, errContains: badRequestErr, contentType: runtime.ContentTypeProtobuf},
```
You don't have any patch test here?

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go:
```diff
@@ -457,6 +458,65 @@ func isDryRun(url *url.URL) bool {
+		return runtime.IgnoreFieldValidation, nil
+	}
+
+	// TODO: Should we blocklist unsupportedContentTypes (just protobuf) rather than allowlisting everything that isn't protobuf?
```
Since this is not a feature that we'll typically get automatically for each new content type (we need to make sure that we can run a strict version of the decoder), I would do an allowlist rather than a blacklist.

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go:
```diff
@@ -116,14 +115,27 @@ func createHandler(r rest.NamedCreater, scope *RequestScope, admit admission.Int
 
+		decoder := scope.Serializer.DecoderToVersion(s.Serializer, scope.HubGroupVersion)
+		trace.Step("About to convert to expected version")
+		obj, gvk, err := decoder.Decode(body, &defaultGVK, original)
```
That answers one of my question above. It doesn't look like ignore and warn are fundamentally different, do we need to keep ignore then? Or if we keep it, does it need to be the default?

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go:
```diff
@@ -116,14 +115,27 @@ func createHandler(r rest.NamedCreater, scope *RequestScope, admit admission.Int
+				return
+			}
+			if validationDirective == runtime.WarnFieldValidation {
+				warning.AddWarning(req.Context(), "", err.Error())
```
I would have expected one warning per unknown field, this looks like it would build a very large warning

apelisse commented on test/integration/apiserver/field_validation_test.go:
```diff
@@ -0,0 +1,681 @@
+
+// smpTestSetup applies an object that will later be patched
+// in the actual test/benchmark.
+func smpTestSetup(t testing.TB, client clientset.Interface) {
```
I'm not a big fan of this function for tests in general

apelisse commented on test/integration/apiserver/field_validation_test.go:
```diff
@@ -0,0 +1,681 @@
+// smpTestSetup applies an object that will later be patched
+// in the actual test/benchmark.
+func smpTestSetup(t testing.TB, client clientset.Interface) {
+	bodyBase, err := os.ReadFile("./testdata/deploy-small.json")
```
I'd also ideally avoid having files read from disk for tests, since it makes it harder to understand what the test does.

apelisse commented on test/integration/apiserver/field_validation_test.go:
```diff
@@ -0,0 +1,681 @@
+		t.Fatalf("failed to read file: %v", err)
+	}
+	putBody := []byte(fmt.Sprintf(string(putBytes), deployName))
+	var testcases = []struct {
+		name         string
+		opts         metav1.UpdateOptions
+		errContains  string
+		warnContains string
```
For this test (and another), I don't know if I enjoy the value of table-driven testing here, as I feel this obfuscates what the test does for very limited code re-use.

kevindelgado commented:
I spent all of today focused on the SMP unstructured converter because that is where I felt the least comfortable and wanted to get some resolution on first. 

In addressing your feedback @apelisse , I think I finally understood the "pre-flattening" algo we had discussed before that I was previously failing to grok. I've updated the converter to use this now which gets rid of the wonky "fields" field we had to add to the various method signatures that you had commented on (and clears up some of the other concerns you had).

On the performance side , the way I had implemented converter strict validation was about 300% slower than what's currently in master "no validation" (and Ignore also suffered the penalty because it was doing the unknown fields check and just not returning the error). Now, I'm seeing roughly parity between strict validation and the old "no validation" (actually, strict appears to be slightly _faster_ and _less_ memory (which is weird and I need to look into it on monday).

One thing I did not implement was the "full path to unknown field" that we discussed offline. I noticed that the json deserializer also does not indicate the full path, so I wasn't sure if we actually wanted to do that here in SMP. I think it's easy enough to add (although I think it will require changing the method signatures to store the path down successive recursive calls of `fromUnstructured`), but wanted to hold off until we talk to Jordan and agree we actually want to preserve the full path in the errors.

Anyways, I know this is a lot of text, so let's sync on Monday. If you want, feel free to take another look at `converter.go`, otherwise I will address the rest of your feedback on Monday.
